File: dotnet/Atlas.Averbacao/Security/ReplayCache.cs
namespace Atlas.Averbacao.Security;

public sealed class ReplayCache
{
    private readonly Dictionary<string, long> _seen = new();
    private readonly long _ttlMs;
    private readonly object _gate = new();

    public ReplayCache(long ttlMs = 60_000) => _ttlMs = ttlMs;

    public bool Claim(string nonce, long? nowMs = null)
    {
        var now = nowMs ?? DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        lock (_gate)
        {
            foreach (var expired in _seen.Where(kv => kv.Value <= now).Select(kv => kv.Key).ToList())
                _seen.Remove(expired);
            if (_seen.ContainsKey(nonce)) return false;
            _seen[nonce] = now + _ttlMs;
            return true;
        }
    }
}

File: java/src/main/java/br/com/atlas/averbacao/security/ReplayCache.java
package br.com.atlas.averbacao.security;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import org.springframework.stereotype.Component;

@Component
public class ReplayCache {
    private final Map<String, Long> seen = new ConcurrentHashMap<>();
    private final long ttlMs;

    public ReplayCache() { this(60_000); }
    public ReplayCache(long ttlMs) { this.ttlMs = ttlMs; }

    public boolean claim(String nonce) {
        long now = System.currentTimeMillis();
        seen.entrySet().removeIf(e -> e.getValue() <= now);
        if (seen.containsKey(nonce)) return false;
        seen.put(nonce, now + ttlMs);
        return true;
    }
}
